WordPress/Classes/ViewRelated/Pages/PageListViewController.swift

```diff
@@ -210,17 +233,55 @@ final class PageListViewController: AbstractPostListViewController, UIViewContro
     override func updateAndPerformFetchRequest() {
         super.updateAndPerformFetchRequest()
 
-        reloadPages()
+        Task {
+            await reloadPagesAndUI()
+        }
     }
 
-    private func reloadPages() {
+    @MainActor
+    private func reloadPagesAndUI() async {
         let status = filterSettings.currentPostListFilter().filterType
         let pages = (fetchResultsController.fetchedObjects ?? []) as! [Page]
+
         if status == .published {
-            self.pages = pages.setHomePageFirst().hierarchySort()
+            let coreDataStack = ContextManager.shared
+            let pageIDs = pages.map { TaggedManagedObjectID($0) }
+
+            do {
+                self.pages = try await buildPageTree(pageIDs: pageIDs)
+                    .hierarchyList(in: coreDataStack.mainContext)
+            } catch {
+                DDLogError("Failed to reload published pages: \(error)")
+            }
         } else {
             self.pages = pages
         }
+
+        tableView.reloadData()
+        refreshResults()
+    }
+
+    /// Build page hierachy in background, which should not take long (less than 2 seconds for 6000+ pages).
```

Comment:
_Show off_ 😄

Reply:
😅 Taking 2 seconds to sort 6000 items is not really fast though...It's much faster than what we have at the moment, but still not fast enough. I'm being lazy here by putting them into a background thread. The comment here is mainly pointing out, the sorting isn't fast enough to take place in the main thread, but is fast enough to put in the background thread without worrying it keep going forever.